src/main/resources/com/google/api/codegen/java/grpc_stub.snip

```diff
@@ -56,7 +56,35 @@
               {@streamingTypeEnum(callable.grpcStreamingType)},
               "{@callable.fullServiceName}/{@callable.protoMethodName}",
               io.grpc.protobuf.ProtoUtils.marshaller({@callable.requestTypeName}.getDefaultInstance()),
-              io.grpc.protobuf.ProtoUtils.marshaller({@callable.responseTypeName}.getDefaultInstance())));
+              io.grpc.protobuf.ProtoUtils.marshaller({@callable.responseTypeName}.getDefaultInstance())),
+          {@requestParamsExtractor(callable)});
+@end
+
+@private requestParamsExtractor(callable)
+  @if callable.hasHeaderRequestParams
+    new RequestParamsExtractor<{@callable.requestTypeName}>() {
+      @@Override
+      public Map<String, String> extract({@callable.requestTypeName} request) {
+        ImmutableMap.Builder<String, String> params = ImmutableMap.builder();
+        {@requestParamExtractor(callable.headerRequestParams)}
+        return params.build();
+      }
+    }
+  @else
+    EmptyRequestParamsExtractor.<{@callable.requestTypeName}>of()
+  @end
```

Comment:
I would prefer for the entry points that don't need a params extractor to not pass in this parameter at all. Otherwise, a *lot* of files will be touched in your regeneration (as you can see in the baseline diff), which we could avoid touching.

Reply:
This is by design. Yes they will be touched and will explicitly say that they don't need parameter extraction. This extraction thing is very close to grpc world (i.e. it would be in grpc spec if changing grpc spec was not so difficult). This is why I try to keep this implementation as close to grpc as possible (i.e. changes are in `DirectCallable`) and explicit (a grpc call should have an explicit name, parameter and header parameters).

Reply:
I understand and agree that the implementation should be close to grpc, but when it comes to codegen, there are some practical matters that come to play. If you don't pass a parameter when there is no extraction, then actually *no* files in google-cloud-java need to be touched right now (since this feature isn't configured for any current APIs). Additionally, I am going to change the interface soon anyway, and I don't want to have "regeneration whiplash" (widescale codegen change, then another one right away).

Reply:
Ok, changed.